flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java

```diff
@@ -342,6 +342,10 @@ private void advanceLastUsingCheckpointID(long checkpointID) {
         }
 
         private void preventDiscardingCreatedCheckpoint() {
+            // Changed from false to true when a newer checkpoint starts reusing this state entry
+            // after recovery. This is to delay discarding the initial checkpoint until all of its
```

Comment:
nit: This is to delay discarding the ~initial~ checkpoint until all of its state (both shared and private) is not used.

The **initial** checkpoint is handled by [`restoredCheckpointSharingStrategies`](https://github.com/apache/flink/pull/20313/commits/34fa6de59b115f960e738b146db4b659287f4a8c#diff-87db13d9f5e3cf08a68052f9eba6c97ab22572fb5b5ccb0cabff7104efb21cebR412).